### ISSUE TITLE
Chat - Users that are not members of the room are suggested when type "@"

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#24436]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to a group chat or #Room that has several members (>6)
2. Enter "@" in the compose box
3. Observe the list of user suggestions

## Expected Result:
5 suggestions should be shown at once inside the mention menu that are members of the room. As more characters are typed, if member of the room contains those characters, they'll always show at top.  Any suggestions after/below that can use our existing search logic (because it's intended to be able to mention anyone in a room, we just want to prioritize showing members first)

## Actual Result:
Users that are not members of the room are suggested when type "@"

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] Windows / Chrome 
- [ ] MacOS / Desktop

**Version Number:** 1.3.53.1

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/78819774/e6712f4b-0489-4780-a466-ac3603547939

**Expensify/Expensify Issue URL:**
 
**Issue reported by:**  Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0106156f4c26122cd1</li>
        <li>Upwork Job ID: 1720587630725419008</li>
        <li>Last Price Increase: 2023-11-03</li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
I'm OOO this week, will review next week once back

**Comment by mallenexpensify** [comment]
Dropped in #expensify-open-source
https://expensify.slack.com/archives/C01GTK53T8Q/p1695117107964409

Applause reported the above, I updated it from Bug to New Feature.  I'd like to move the issue folder since it makes sense to me to that the members of a room should show first in the recommendations (this happens to me a lot in a private room, when I try to tag @drew the top three recommendations are Andrew and none are members of the room)
If anyone has any feedback or context on how we display our recommendations (and if this can be External) can you please comment on the issue or in :thread: ?  (also if you don't agree we implement the new feature, or to not do it now)

**Comment by mallenexpensify** [comment]
Will look into when I have time

**Comment by mallenexpensify** [comment]
Need to prioritize soon as I think it's a useful feature request

**Comment by mallenexpensify** [comment]
Now wondering if this is something that might be part of a future wave.   hmmm... 
Also, it's likely not a big issue for newer users, since they won't have a lot of cross-company chats, most will be internal on one domain

**Comment by mallenexpensify** [comment]
Soon, I hope.  Ran into this issue recently. @cubuspl42 , at a high level, do you know how hard this might be and/or if it can be worked on externally?  The logic should be pretty straight forward, I think. We might want to consider rooms and workspaces at the same time though.  ie. 
I can break down examples if the above isn't clear.   Thx

**Comment by cubuspl42** [comment]
@spcheema Thanks for linking that issue. I believe its resolution also applies to us, though...

Per [this Slack message](https://expensify.slack.com/archives/C02NK2DQWUX/p1691080525453239?thread_ts=1691075159.942679&cid=C02NK2DQWUX):

> That is not a bug. You should be able to mention anyone anywhere

I think this issue should be closed.

**Comment by mallenexpensify** [comment]
I think there's some confusion here.  
> You should be able to mention anyone anywhere

If I'm in the same workspace or room with someone and I start typing their name (and, let's say I've chatted with them 100 times), there name should show before someone who's not in the room (and, lets say I've only chatted with them once). 

So... maybe the logic would be. 
For Room
1. If a user is part of that room, always show there name first in a search (where all else is equal.  ie. `tim` who is in the room would always shows before any `tim` who is not)
2. If a user is part of a workspace, always show there name first in a search after #1 above (where all else is equal.  ie. `tim` who is in the room would always show first, `tim` in the same workspace as the room would show second and  any other `tim` would show after (ideally with the ranking being were the `tim` you chatted with most would show at top. 

Before I address in Slack to get more 👀, does anyone have feedback on the above? It's quite possible this was discussed before.  Also... for the above, are there any known issues that would keep us from implementing the logic to work?

**Comment by cubuspl42** [comment]
@mallenexpensify I'm not arguing with anything you say per se, just in the spirit of the recent announcements, the current focus is on clearing the backlog and solving issues blocking the customers from moving from OldDot

**Comment by cubuspl42** [comment]
So I'm suggesting closing with resolution "Won't fix _for now_" 🤷‍♂️

**Comment by mallenexpensify** [comment]
Thanks for your thoughts @cubuspl42 

I can't think of a reason or benefit to showing a user as an option to tag in a room, if they're not already a member, above showing a user who is a member of a room.  Here's a real-world example.  
Users in a room
<img width="434" alt="image" src="https://github.com/Expensify/App/assets/22508304/1acd38e9-7883-4785-83e9-9ded20a27b23">

When, in the room, I start typing `@a` this is what I see
<img width="798" alt="image" src="https://github.com/Expensify/App/assets/22508304/8bfee95b-5874-4901-afe6-2c61328a53e4">

What's the benefit of showing non-members in this instance? 

I think the update would make it easier/better for new and existing users in a room.

**Comment by cubuspl42** [comment]
🎀 👀 🎀

**Comment by cubuspl42** [comment]
@youssef-lr 

We have a proposal and some ongoing discussion, but we also have a quote from an internal engineer that this is _not a bug_ (or at least not a bug worth prioritizing).

I believe that the call whether this _should_ or _shouldn't_ be worked on is not up to me here.

**Comment by mallenexpensify** [comment]
@cubuspl42 I moved to to `New Feature` and removed the `Bug` label, so it's less priority than a bug but I think we should build the new feature.  Here's another 'real world' example.  Each time (which has been dozens) I have to either navigate to the user to click their name or scroll down multiple times then tap `return`.  

The person I want to tag is Drew, and they're the only one's in the room, so the other recommended users aren't helpful.  
<img width="436" alt="image" src="https://github.com/Expensify/App/assets/22508304/334af1c4-8bc8-4764-bc7f-0cbb433ec7f4">


> Per [this Slack message](https://expensify.slack.com/archives/C02NK2DQWUX/p1691080525453239?thread_ts=1691075159.942679&cid=C02NK2DQWUX):

That is not a bug. You should be able to mention anyone anywhere

Correct, you should be able to mention/type any email, we shouldn't have users that aren't in the room be the first that are surfaced/recommended though.

**Comment by mallenexpensify** [comment]
@youssef-lr can you please review this when you have a minute (it's not high priority, it's a feature request). 
My _dream_ is that, if you agree with the expected behavior below, contributors can do most-all of the work

<img width="835" alt="image" src="https://github.com/Expensify/App/assets/22508304/d4036791-cda2-4acc-abcf-153761048e7f">

**Comment by mallenexpensify** [comment]
I'm working on search results organization/prioritization and this _might_ factor in, should 
know more soon.

**Comment by youssef-lr** [comment]
@mallenexpensify sorry I was OOO when this was created and I've been focusing on high priority issues. I will take a look today and get back to you!

**Comment by youssef-lr** [comment]
@mallenexpensify [this](https://github.com/Expensify/App/issues/24436#issuecomment-1821879734) looks good to me!

**Comment by mallenexpensify** [comment]
Thanks @youssef-lr , gonna take me a lil bit to get back to this cuz wanna be sure we're addressing holistically with search results ordering, which I'm going to work on today/tomorrow.

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by mallenexpensify** [comment]
Shared in a bug report in #expensify-bugs today, request was for workspace returns though, not specifically rooms
https://expensify.slack.com/archives/C049HHMV9SM/p1705599127821279?thread_ts=1705576735.390529&cid=C049HHMV9SM

**Comment by mallenexpensify** [comment]
From Mills last week in the link above. 
> I do think we should start prioritizing the search results.

Reckon it's just a matter of time til we address.

**Comment by mallenexpensify** [comment]
Still waiting to prioritize search results to know what Wave/VIP to assign this too, was thinking #vip-vsb but we want to approach holistically and not _just_ fix this one issue.

**Comment by youssef-lr** [comment]
Not overdue.

**Comment by youssef-lr** [comment]
Same.

**Comment by youssef-lr** [comment]
@mallenexpensify have you figured out where this fits and its priority?

**Comment by mallenexpensify** [comment]
Added to #vip-vsb since it doesn't involve money.  It's a feature request so it might not get addressed soon. 
@youssef-lr reported similar functionality for group chats today too, [here](https://expensify.slack.com/archives/C049HHMV9SM/p1710878688977429)

**Comment by mallenexpensify** [comment]
@cubuspl42 @youssef-lr , do you think this feature request can be addressed one-off or should we get more 👀 in #expensify-open-source to make sure any changes that would be made, wouldn't conflict with anything else?

**Comment by spcheema** [comment]
> @mallenexpensify @cubuspl42 it's pretty straight forward to restrict non-members being tagged in a room/workspace.
> 
> I have submitted RCA & solution here while back
> 
> https://github.com/Expensify/App/issues/22596#issuecomment-1629904781
> 
> 
> It can be done externally since it requires no backend change.

Hi @mallenexpensify @cubuspl42 

I did submitted proposal long ago but then issues decided not to proceed with.

If we are going ahead then I'll update proposal accordingly.

**Comment by mallenexpensify** [comment]
Bumping to weekly while we await prioritization and/or a holistic plan.

**Comment by mallenexpensify** [comment]
Checking on [here](https://expensify.slack.com/archives/C049HHMV9SM/p1713322833240619?thread_ts=1711460237.297499&cid=C049HHMV9SM)

**Comment by mallenexpensify** [comment]
Checking for a Search v2 link
https://expensify.slack.com/archives/C049HHMV9SM/p1716170420813589?thread_ts=1711460237.297499&cid=C049HHMV9SM

**Comment by mallenexpensify** [comment]
Going to close for now, since I think/hope we'll be addressing this more _holistically_ soon, as part of Search v2.
